src/Raven.Server/Background/BackgroundWorkBase.cs

```diff
@@ -84,6 +91,10 @@ protected async Task WaitOrThrowOperationCanceled(TimeSpan time)
 
                 // if cancellation requested then it will throw TaskCancelledException and we stop the work
                 await TimeoutManager.WaitFor(time, CancellationToken).ConfigureAwait(false); 
+                if(CancellationToken.IsCancellationRequested)
```

Comment:
Shouldn't the `WaitFor` already throw that?

Reply:
It doesn't I have checked we will get stuck in an infinite loop

Reply:
We also need to update the `WaitFor`, then.

Reply:
I think it would change its current behavior and it is risky, no?

Reply:
I expected that to fail, and not hang, if the cancellation token is cancelled.

Reply:
what is hanging is the expired document cleaner logic, not the timing manager.
But if you pass a default cancelation token to the timing manager it will hang so it is a problem with the usage, not the underlining logic.
@maximburyak will open an issue to review usages

Reply:
We'll address that in a different issue: https://issues.hibernatingrhinos.com/issue/RavenDB-13849